File: plugin/ios/Plugin/SimpleBPETokenizer.swift
import Foundation

/// Simple BPE Tokenizer for SentencePiece vocab
/// Implements greedy longest-match tokenization
class SimpleBPETokenizer {
    private var vocab: [String: Int] = [:]
    private var reverseVocab: [Int: String] = [:]
    private var specialTokens: Set<String> = []
    
    // Special token IDs for M2M100/SMALL100
    static let BOS_TOKEN_ID = 0
    static let PAD_TOKEN_ID = 1
    static let EOS_TOKEN_ID = 2
    static let UNK_TOKEN_ID = 3
    
    init(vocab: [String: Int]) {
        self.vocab = vocab
        
        // Build reverse vocab for decoding
        for (token, id) in vocab {
            reverseVocab[id] = token
        }
        
        // Mark special tokens
        specialTokens = ["<s>", "</s>", "<pad>", "<unk>"]
    }
    
    /// Encode text to token IDs using greedy longest-match
    func encode(_ text: String) -> [Int] {
        // SentencePiece uses ▁ (U+2581) as word boundary marker
        // Normalize text: add ▁ at start and replace spaces with ▁
        let normalizedText = "▁" + text.replacingOccurrences(of: " ", with: "▁")
        
        var tokens: [Int] = []
        var i = normalizedText.startIndex
        
        while i < normalizedText.endIndex {
            var found = false
            let remaining = normalizedText.distance(from: i, to: normalizedText.endIndex)
            let maxLen = min(20, remaining)
            
            // Try longest match first
            for len in stride(from: maxLen, through: 1, by: -1) {
                let endIndex = normalizedText.index(i, offsetBy: len)
                let substr = String(normalizedText[i..<endIndex])
                
                if let tokenId = vocab[substr] {
                    tokens.append(tokenId)
                    i = endIndex
                    found = true
                    break
                }
            }
            
            if !found {
                // Unknown character - use <unk> token
                if let unkId = vocab["<unk>"] {
                    tokens.append(unkId)
                }
                i = normalizedText.index(after: i)
            }
        }
        
        return tokens
    }
    
    /// Decode token IDs back to text
    func decode(_ ids: [Int], skipSpecialTokens: Bool = true) -> String {
        var result = ""
        
        for id in ids {
            guard let token = reverseVocab[id] else { continue }
            
            if skipSpecialTokens && specialTokens.contains(token) {
                continue
            }
            
            result += token
        }
        
        // Replace ▁ with space and trim
        return result.replacingOccurrences(of: "▁", with: " ").trimmingCharacters(in: .whitespaces)
    }
    
    /// Decode single token ID to string (for debugging)
    func decodeToken(_ id: Int) -> String {
        return reverseVocab[id] ?? "<unknown>"
    }
    
    var vocabSize: Int {
        return vocab.count
    }
    
    var bosTokenId: Int { Self.BOS_TOKEN_ID }
    var eosTokenId: Int { Self.EOS_TOKEN_ID }
    var padTokenId: Int { Self.PAD_TOKEN_ID }
    var unkTokenId: Int { Self.UNK_TOKEN_ID }
}

File: plugin/android/src/main/java/com/small100onnx/SimpleBPETokenizer.java
package com.small100onnx;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Simple BPE Tokenizer for SentencePiece vocab
 * Implements greedy longest-match tokenization
 */
public class SimpleBPETokenizer {
    private Map<String, Integer> vocab;
    private Map<Integer, String> reverseVocab;
    private Set<String> specialTokens;
    
    // Special token IDs for M2M100/SMALL100
    private static final int BOS_TOKEN_ID = 0;
    private static final int PAD_TOKEN_ID = 1;
    private static final int EOS_TOKEN_ID = 2;
    private static final int UNK_TOKEN_ID = 3;
    
    public SimpleBPETokenizer(Map<String, Integer> vocab) {
        this.vocab = vocab;
        this.reverseVocab = new HashMap<>();
        this.specialTokens = new HashSet<>();
        
        // Build reverse vocab for decoding
        for (Map.Entry<String, Integer> entry : vocab.entrySet()) {
            reverseVocab.put(entry.getValue(), entry.getKey());
        }
        
        // Mark special tokens
        specialTokens.add("<s>");
        specialTokens.add("</s>");
        specialTokens.add("<pad>");
        specialTokens.add("<unk>");
    }
    
    /**
     * Encode text to token IDs using greedy longest-match
     */
    public int[] encode(String text) {
        // SentencePiece uses ▁ (U+2581) as word boundary marker
        // Normalize text: add ▁ at start and replace spaces with ▁
        String normalizedText = "▁" + text.replace(" ", "▁");
        
        List<Integer> tokens = new ArrayList<>();
        int i = 0;
        
        while (i < normalizedText.length()) {
            boolean found = false;
            // Try longest match first (up to 20 chars)
            int maxLen = Math.min(20, normalizedText.length() - i);
            
            for (int len = maxLen; len > 0; len--) {
                String substr = normalizedText.substring(i, i + len);
                Integer tokenId = vocab.get(substr);
                
                if (tokenId != null) {
                    tokens.add(tokenId);
                    i += len;
                    found = true;
                    break;
                }
            }
            
            if (!found) {
                // Unknown character - use <unk> token
                Integer unkId = vocab.get("<unk>");
                if (unkId != null) {
                    tokens.add(unkId);
                }
                i++;
            }
        }
        
        // Convert to int array
        int[] result = new int[tokens.size()];
        for (int j = 0; j < tokens.size(); j++) {
            result[j] = tokens.get(j);
        }
        return result;
    }
    
    /**
     * Decode token IDs back to text
     */
    public String decode(int[] ids, boolean skipSpecialTokens) {
        StringBuilder sb = new StringBuilder();
        
        for (int id : ids) {
            String token = reverseVocab.get(id);
            if (token == null) continue;
            
            if (skipSpecialTokens && specialTokens.contains(token)) {
                continue;
            }
            
            sb.append(token);
        }
        
        // Replace ▁ with space and trim
        String result = sb.toString().replace("▁", " ").trim();
        return result;
    }
    
    /**
     * Decode single token ID to string (for debugging)
     */
    public String decodeToken(int id) {
        String token = reverseVocab.get(id);
        return token != null ? token : "<unknown>";
    }
    
    public int getVocabSize() {
        return vocab.size();
    }
    
    public int getBosTokenId() {
        return BOS_TOKEN_ID;
    }
    
    public int getEosTokenId() {
        return EOS_TOKEN_ID;
    }
    
    public int getPadTokenId() {
        return PAD_TOKEN_ID;
    }
    
    public int getUnkTokenId() {
        return UNK_TOKEN_ID;
    }
}
